### ISSUE TITLE
Migration does not respect transaction

### ISSUE THREAD

**Issue by maxbry** [gobuffalo/pop#678]
In my migration, I have two commands. The first one adds a new column to a table, the second one adds a foreign key:

```
add_column("visitors", "date_id", "int", {"null":true})
add_foreign_key("visitors", "non_existing_column", {"time_intervals": ["id"]}, {
    "name": "pv_ti_fk",
    "on_delete": "restrict",
    "on_update": "restrict",
})
```

As you can see, the column used for the column is incorrect, which makes the migration fail. However, when executing the migration again, it says the column `date_id` already exists. This shows that the first command has been comitted, although the full transaction should have been rolled back.

**Comment by sio4** [comment]
Can you post your package versions? If you are using `pop` within the buffalo app, I guess so, the result of `buffalo info` from the app root will be very helpful. Otherwise, the result of `soda --version` or `buffalo pop --version` could be fine.

Also, please post the full output when you run the command for the first time (with an empty database) and the second result when you run the command again after one failure. The command line could be extended with `-d` debugging option so the command can be executed as `soda migrate up -d` if you use `soda` or `buffalo pop migrate up -d`.

My expectation is that the fizz file will be interpreted by `soda` and the two lines of SQL (for add_column and add_foreign_key) should be executed within one transaction.

**Comment by sio4** [comment]
Ah! I confirmed the behavior but it would be also good to see your versions. :-)

**Comment by sio4** [comment]
I checked the code and I found that we have no special treatment for migration-specific rollback which means it just depends on the behavior of the underlying database system. Some databases support transactional DDL natively, while others do not.

I think implementing it as a workaround (just run `migrate down` for failed migration version by adding a status indicator for migrations) is technically possible, but I am not sure if that feature is really needed.

**Comment by maxbry** [comment]
This is my buffalo version: `INFO[0001] Buffalo version is: v0.17.7`

Running `migrate down` wouldn't be a sufficient solution: Imagine I have three commands in one migration. The downward migration could revert everything, but how does it know what already has been done and what is missing? So we would have the same problem I guess

**Comment by sio4** [comment]
What I understand from your first comment is that you expected that the successful command, `add_column()`, could be reverted back when the second command `add_foreign_key()` failed. Is it correct?

Then the migration down for failed file will work (revert everything in the file) if we just ignore the execution failure (in this case failure on the  `drop_foreign_key()` since the FK generation failed before) for down and just run `drop_column()` after failing of dropping fk, or using an additional status indicator as I already mentioned. (currently, we only store the succeeded migrations.) Anyway, you are also correct since we cannot guarantee if the developer who wrote the wrong migration commands for up wrote perfect migration commands for down. It cannot be perfect.

So yeah, I don't think that feature is really needed, TBH, I am in the opposite part of it since:
- Basically, it is not an issue of pop but a limitation of the underlying database.
- Migration should be written by developers with valid values/syntax and this is the responsibility of developers.
- This issue could affect only development time and easily be prevented by developers.
- I don't think this issue could affect the end user's interaction with the deployed application.

**Comment by maxbry** [comment]
If the problem is on the database side and not on pop, then I agree with your points.

**Comment by sio4** [comment]
> If the problem is on the database side and not on pop, then I agree with your points.

Yeah, you can search for "transactional DDL" for more information.

e.g. https://www.datasunrise.com/blog/professional-info/how-popular-rdbmss-deal-with-ddl-commands-in-transactions/

**Comment by maxbry** [comment]
Yes, I found similar information. Since my issue then is no `pop` issue, I will close it. Thank you for your comments though!
